### ISSUE TITLE
question about reproducing

### ISSUE THREAD

**Issue by jinxxo-j** [lizr16/OMAD#2]
Hi, thank you for releasing the OMAD code. I am trying to reproduce the MaMuJoCo results in Table 3 before evaluating an extension of OMAD, and I would appreciate clarification on the reward convention and exact environment setup.

Preliminary reproduction results

I used the task-specific hyperparameters reported in Tables 1 and 2.

The results below are the maximum evaluation returns observed during training. Except for Cooperative Navigation with three agents, they currently use only seed 0 and should therefore be considered preliminary.

<img width="781" height="254" alt="Image" src="https://github.com/user-attachments/assets/af64686e-d12e-4f8c-917c-bb0a734741f1" />

Most Ant, Walker2d, and Swimmer results are reasonably close to the reported values after accounting for the number of agents. However, the reproduced MPE results are substantially higher than those in the paper, while HalfCheetah 6x1 remains substantially lower.

This task-dependent pattern makes me suspect that the reward convention, environment version, observation configuration, or evaluation protocol may differ between my setup and the setup used for Table 3.

1. MaMuJoCo reward aggregation

mamujoco_v1 returns the same Gymnasium environment reward to every agent. The released OMAD wrapper then sums the rewards of all trainable agents. Therefore, for an N-agent MaMuJoCo task, the scalar stored in the replay buffer and logged by the wrapper appears to be:

`N × environment reward`

For example:

- Ant 2x4d: raw return 15444.23, which becomes 7722.12 after dividing by 2.

- HalfCheetah 6x1: raw return 45603.68, which becomes 7600.61 after dividing by 6.

- Walker2d 2x3: raw return 15624.05, which becomes 7812.03 after dividing by 2.

The normalized values are much closer to the scale reported in Table 3 than the raw values produced by the released implementation.

Could you please clarify which convention was used for the paper?

Train using the sum of duplicated agent rewards, but divide evaluation returns by N only for plotting and reporting.

Use one shared environment reward, equivalently the mean across agents, for both training and evaluation.

Use another convention.

This distinction seems important because it changes not only the reported return but also the training target and its scale relative to the distributional critic’s V_min and V_max.

2. Exact MaMuJoCo versions

The released requirements.txt specifies:

`gymnasium>=0.29.0
pettingzoo>=1.24.0`

However, it does not list or pin gymnasium-robotics or mujoco, while the implementation imports gymnasium_robotics.mamujoco_v1.

My current environment is:

- Python 3.11.15

- Gymnasium 1.2.1

- Gymnasium-Robotics 1.3.1

- PettingZoo 1.25.0

- MuJoCo 2.3.7

- JAX/JAXLIB 0.4.33

Could you share the exact versions used for the paper? A lock file, pip freeze, or Conda environment export would be especially helpful.

3. MaMuJoCo observation settings

Could you also confirm whether the paper used the default mamujoco_v1.parallel_env(...) settings, including:

- agent_obsk=1

- Default local and global observation categories

- Ant observations including cfrc_ext

- The Walker2d v1 global-node fix

- A maximum episode length of 1,000 steps

With my current environment, the per-agent observation dimensions are:

`Ant 2x4 and 2x4d: 63
Walker2d 2x3: 12`

Do these dimensions match the setup used for Table 3?

4. Evaluation protocol

The paper states that Table 3 reports the maximum average episode return over five seeds. Could you please clarify:

- The five seed values

- The number of evaluation episodes per checkpoint

- Whether the maximum was selected from the mean learning curve at a common checkpoint across seeds

- Whether the reported standard deviation was calculated across the five seed returns at that checkpoint

Thank you for your help. Once these details are clarified, I plan to run the complete five-seed comparison using the confirmed protocol.

**Comment by jinxxo-j** [comment]
> Hi, thank you for releasing the OMAD code. I am trying to reproduce the MaMuJoCo results in Table 3 before evaluating an extension of OMAD, and I would appreciate clarification on the reward convention and exact environment setup.
> 
> Preliminary reproduction results
> 
> I used the task-specific hyperparameters reported in Tables 1 and 2.
> 
> The results below are the maximum evaluation returns observed during training. Except for Cooperative Navigation with three agents, they currently use only seed 0 and should therefore be considered preliminary.
> 
> <img alt="Image" width="781" height="254" src="https://private-user-images.githubusercontent.com/126158541/641998635-af64686e-d12e-4f8c-917c-bb0a734741f1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODc4MzY1NDUsIm5iZiI6MTc4NzgzNjI0NSwicGF0aCI6Ii8xMjYxNTg1NDEvNjQxOTk4NjM1LWFmNjQ2ODZlLWQxMmUtNGY4Yy05MTdjLWJiMGE3MzQ3NDFmMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwODI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDgyN1QxMzEwNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MTljYmRjMTQ1MmM3MzhiYThiMzA1ZGIxNmJlODQ2NzhhNzcyMWU2MDA1MzY4ZDVlMGZlYzYzYTM0NGU1NTczJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.EitLBZuuEC2vplWKKY888yA6maPJK2Bobq8NVJFzomw">
> Most Ant, Walker2d, and Swimmer results are reasonably close to the reported values after accounting for the number of agents. However, the reproduced MPE results are substantially higher than those in the paper, while HalfCheetah 6x1 remains substantially lower.
> 
> This task-dependent pattern makes me suspect that the reward convention, environment version, observation configuration, or evaluation protocol may differ between my setup and the setup used for Table 3.
> 
> 1. MaMuJoCo reward aggregation
> 
> mamujoco_v1 returns the same Gymnasium environment reward to every agent. The released OMAD wrapper then sums the rewards of all trainable agents. Therefore, for an N-agent MaMuJoCo task, the scalar stored in the replay buffer and logged by the wrapper appears to be:
> 
> `N × environment reward`
> 
> For example:
> 
> * Ant 2x4d: raw return 15444.23, which becomes 7722.12 after dividing by 2.
> * HalfCheetah 6x1: raw return 45603.68, which becomes 7600.61 after dividing by 6.
> * Walker2d 2x3: raw return 15624.05, which becomes 7812.03 after dividing by 2.
> 
> The normalized values are much closer to the scale reported in Table 3 than the raw values produced by the released implementation.
> 
> Could you please clarify which convention was used for the paper?
> 
> Train using the sum of duplicated agent rewards, but divide evaluation returns by N only for plotting and reporting.
> 
> Use one shared environment reward, equivalently the mean across agents, for both training and evaluation.
> 
> Use another convention.
> 
> This distinction seems important because it changes not only the reported return but also the training target and its scale relative to the distributional critic’s V_min and V_max.
> 
> 2. Exact MaMuJoCo versions
> 
> The released requirements.txt specifies:
> 
> `gymnasium>=0.29.0 pettingzoo>=1.24.0`
> 
> However, it does not list or pin gymnasium-robotics or mujoco, while the implementation imports gymnasium_robotics.mamujoco_v1.
> 
> My current environment is:
> 
> * Python 3.11.15
> * Gymnasium 1.2.1
> * Gymnasium-Robotics 1.3.1
> * PettingZoo 1.25.0
> * MuJoCo 2.3.7
> * JAX/JAXLIB 0.4.33
> 
> Could you share the exact versions used for the paper? A lock file, pip freeze, or Conda environment export would be especially helpful.
> 
> 3. MaMuJoCo observation settings
> 
> Could you also confirm whether the paper used the default mamujoco_v1.parallel_env(...) settings, including:
> 
> * agent_obsk=1
> * Default local and global observation categories
> * Ant observations including cfrc_ext
> * The Walker2d v1 global-node fix
> * A maximum episode length of 1,000 steps
> 
> With my current environment, the per-agent observation dimensions are:
> 
> `Ant 2x4 and 2x4d: 63 Walker2d 2x3: 12`
> 
> Do these dimensions match the setup used for Table 3?
> 
> 4. Evaluation protocol
> 
> The paper states that Table 3 reports the maximum average episode return over five seeds. Could you please clarify:
> 
> * The five seed values
> * The number of evaluation episodes per checkpoint
> * Whether the maximum was selected from the mean learning curve at a common checkpoint across seeds
> * Whether the reported standard deviation was calculated across the five seed returns at that checkpoint
> 
> Thank you for your help. Once these details are clarified, I plan to run the complete five-seed comparison using the confirmed protocol.

4. Deterministic evaluation and diffusion noise

The evaluation callback calls:

model.predict(observation, deterministic=True)

However, the diffusion policy’s _predict() method does not appear to use the deterministic argument. It resets the JAX noise key for every prediction, samples a new Gaussian prior, and adds stochastic noise during the denoising integration steps.

Therefore, evaluation appears to remain stochastic even when deterministic=True.

Could you please clarify how diffusion noise was handled when producing the learning curves and Table 3?

Was a fixed initial noise sample or fixed random key used during evaluation?

Was the denoising process made deterministic by removing the per-step stochastic noise?

Or were actions sampled stochastically with fresh noise, as in the released implementation?

This may materially affect both the evaluation variance and the reported maximum return across checkpoints.

**Comment by m0feng** [comment]
Could I take a look at your reward curve for Walker2d 2x3? I'm reproducing the results, but the evaluation reward I'm getting has quite a lot of fluctuation.
I'm using the JaxMarl environment, and for the same benchmark, the performance drops significantly.

**Comment by jinxxo-j** [comment]
Sure. I have attached the training and evaluation curves from my Walker2d 2x3 experiment.

<img width="1143" height="716" alt="Image" src="https://github.com/user-attachments/assets/d7f5548f-10c9-41b7-ae47-1b2e9978a175" />
<img width="1141" height="366" alt="Image" src="https://github.com/user-attachments/assets/302a5698-9bca-467c-90d0-c8ba9c51c38c" />
